python/kaon_physics.py

```diff
@@ -0,0 +1,109 @@
+
+class KaonPhysics():
+    """Parameters that determine the physics of kaons in the simulation
+
+    Parameters
+    ----------
+    kplus_branching_ratios : list[float]
+    kminus_branching_ratios : list[float]
+    k0l_branching_ratios : list[float]
+    k0s_branching_ratios : list[float]
+        List of the desired branching ratios for kaons. The entries
+        correspond go the following processes
+
+        For charged kaons
+        - 0: K -> mu + v
+        - 1: K -> pi + pi0
+        - 2: K -> 3 pi
+        - 3: K -> pi + 2 pi0
+        - 4: K -> pi0 + e + v
+        - 5: K -> pi0 + mu + v
+
+        For Klong
+        - 0: K -> 3 pi0
+        - 1: K -> pi0 + pi+ + pi-
+        - 2: K -> pi- + e+ + v
+        - 3: K -> pi+ + e- + v
+        - 4: K -> pi- + mu+ + v
+        - 5: K -> pi+ + mu- + v
+
+        For Kshort:
+        - 0: K -> pi+ + pi-
+        - 1: K -> 2 pi0
+
+        See the sources for Geant4's definitions and branching ratios
+        (G4KaonMinus.cc, G4KaonPlus.cc, KaonZeroLong.cc, KaonZeroShort.cc). The
+        default branching ratios correspond to the Geant4 defaults.
+
+    kplus_lifetime_factor : float
+    kminus_lifetime_factor : float
+    k0l_lifetime_factor : float
+    k0s_lifetime_factor : float
+
+        Multiplicative factor to scale the lifetime of kaons by. Default is to
+        scale by 1 (no scaling)
+
+    """
+    def __init__(self):
+        self.kplus_branching_ratios = [
+            0.6355,  # K^+ -> mu^+ + nu_mu
+            0.2066,  # K^+ -> pi^+ + pi^0
+            0.0559,  # K^+ -> pi^+ + pi^- + pi^+
+            0.01761, # K^+ -> pi^+ + pi^0 + pi^0
+            0.0507,  # K^+ -> pi^0 + e^+ + nu_e
+            0.0335,  # K^+ -> pi^0 + mu^+ + nu_mu
+        ]
+        self.kminus_branching_ratios = [
+            0.6355,  # K^- -> mu^- + anti_nu_mu
+            0.2066,  # K^- -> pi^- + pi^0
+            0.0559,  # K^- -> pi^- + pi^+ + pi^-
+            0.01761, # K-+ -> pi^- + pi^0 + pi^0
+            0.0507,  # K-+ -> pi^0 + e^- + anti_nu_e
+            0.0335,  # K-+ -> pi^0 + mu^- + anti_nu_mu
+        ]
+        self.k0l_branching_ratios = [
+            0.1952, # K^0_L -> pi^0 + pi^0 + pi^0
+            0.1254, # K^0_L -> pi^0 + pi^+ + pi^-
+            0.2027, # K^0_L -> pi^- + e^+ + nu_e
+            0.2027, # K^0_L -> pi^+ + e^- + anti_nu_e
+            0.1352, # K^0_L -> pi^- + mu^+ + nu_mu
+            0.1352, # K^0_L -> pi^+ + mu^- + anti_nu_mu
+        ]
+        self.k0s_branching_ratios = [
+            0.6920, # K^0_S -> pi^+ + pi^-
+            0.3069, # K^0_S -> pi^0 + pi^0
+        ]
```

Comment:
Do the branching ratios need to add up to `1`? What happens if they don't?

This kinda naturally leads me to: should we add a check for this? We could customize `__setattr__` to do some checking on the values passed.

```python

def __setattr__(self, key, value):
    if 'branching_ratios' in key:
        super().__setattr__(key, value)
    else:
        if not isinstance(value, list):
            raise TypeError('Values of branching ratios need to be lists')
        if sum(value) != 1.:
            raise ValueError(f'Total of branching ratios in {key} not equal to one.')
        super().__setattr__(key, value)
```

Reply:
So the first thing I did was to add a runtime check that ensured that they do add up to one. Turns out the default geant4 ones don't actually add up to one... (See https://apc.u-paris.fr/~franco/g4doxy/html/G4KaonMinus_8cc-source.html for example). I haven't checked explicitly, but I _suspect_ that Geant4 mostly cares about the relative rates. 

I was considering adding a check that "hey this is very different from one" though, could still be useful

Reply:
```
        if 'branching_ratios' in key:
            super().__setattr__(key, value)
        else:
```

My python ignorance showing here, but I'm guessing this is supposed to be if branching ratios is not in key?

Reply:
Something like this?

```python
    def __setattr__(self, key, value):
        """Ensure that attempts to set the branching ratios give a total of
        something close to 1 and that the lifetime factors are positive.

        Note that Geant4's defaults are not exactly equal to one, but something
        significantly different from one is likely to be a mistake.

        """
        if 'branching_ratios' in key:
            # super().__setattr__(key, value)
            if not isinstance(value, list):
                raise TypeError(f'Values of branching ratios ({key}) need to be lists')
            if abs(sum(value) - 1) > 0.95:
                raise ValueError(f'Total of branching ratios in {key} significantly different from one, was {value}.')
            super().__setattr__(key, value)
        elif 'lifetime' in key:
            if not isinstance(value, float):
                raise TypeError(f'Lifetime parameter ({key}) needs to be floating-point')
            if value < 0:
                raise ValueError(f'Lifetime parameter ({key}) needs to be positive')
            pass
        else:
            super().__setattr__(key, value)
```

Reply:
I don't mind adding it and if it saves someone in the future from wondering, it might be worth it. But I meant that the check was inverted in your original version (it would do the checks if 'branching_ratios' was not part of the type)

Reply:
Fair point you did notice my own typo.

(As a side note, this slightly connects to a Framework-level major refactor where we could require users to define parameters and their types during class construction and then have a custom `__setattr__` to check everything. I haven't implemented this yet because it would require re-writing all of the downstream config classes, but it is fun to think about removing all typo and type-related issues. https://github.com/LDMX-Software/fire/issues/35)